lib/rack/http_streaming_response.rb

```diff
@@ -18,60 +21,63 @@ def body
     end
 
     def code
-      response.code.to_i
+      response.code.to_i.tap do |response_code|
+        close_connection if STATUSES_WITH_NO_ENTITY_BODY.include?(response_code)
+      end
     end
     # #status is deprecated
     alias_method :status, :code
 
     def headers
-      h = Utils::HeaderHash.new
-
-      response.to_hash.each do |k, v|
-        h[k] = v
+      Utils::HeaderHash.new.tap do |h|
+        response.to_hash.each { |k, v| h[k] = v }
       end
-
-      h
     end
 
     # Can be called only once!
     def each(&block)
+      return if connection_closed
+
       response.read_body(&block)
     ensure
+      return if connection_closed
+
```

Comment:
Perhaps `close_connection` could check `connection_closed` internally so that it could be reused here as well.

Reply:
Yeah, agree!